### PR TITLE
scheduler: support range for schedulers

### PR DESCRIPTION
**Pull request by rleungx**
### What problem does this PR solve? 
Currently, the schedulers don't support the key range. e.g. Given a key range, the scheduler only works in this range. And thus to make the scheduler more flexible, we need to support range for schedulers.

### What is changed and how it works?
This PR supports range for schedulers. It should be reviewed after 

- [x] #1783 is merged.

### Check List 

Tests 

 - Unit test

**edit:** 

- [ ] add more unit tests

### CODE REVIEW

File: server/core/basic_cluster.go [hunk 1]
```diff
@@ -308,9 +308,9 @@ func (bc *BasicCluster) Length() int {
 
 // RegionSetInformer provides access to a shared informer of regions.
 type RegionSetInformer interface {
-	RandFollowerRegion(storeID uint64, opts ...RegionOption) *RegionInfo
-	RandLeaderRegion(storeID uint64, opts ...RegionOption) *RegionInfo
-	RandPendingRegion(storeID uint64, opts ...RegionOption) *RegionInfo
+	RandFollowerRegion(storeID uint64, startKey, endKey []byte, opts ...RegionOption) *RegionInfo
```
[the remark this answers is not in the document]
Comment (nolouch) [reply server/core/basic_cluster.go]: May we need to support multiple ranges?

File: server/schedulers/shuffle_hot_region.go [hunk 2]
```diff
@@ -43,6 +45,7 @@ func init() {
 				}
 				conf.Limit = limit
 			}
+			conf.Name = shuffleHotRegionName
```
[the remark this answers is not in the document]
Comment (rleungx) [reply server/schedulers/shuffle_hot_region.go]: I'm not sure if the hot region related schedulers should support the range. /cc @nolouch

File: server/core/region.go [hunk 3]
```diff
@@ -809,9 +810,11 @@ type RegionsContainer interface {
 	RandomRegion(startKey, endKey []byte) *RegionInfo
 }
 
-func randRegion(regions RegionsContainer, opts ...RegionOption) *RegionInfo {
+func randRegion(regions RegionsContainer, ranges []KeyRange, opts ...RegionOption) *RegionInfo {
```
[the remark this answers is not in the document]
Comment (rleungx) [reply server/core/region.go]: Can you explain more?

File: server/schedulers/utils.go [hunk 4]
```diff
@@ -148,3 +149,25 @@ const (
...
+	for len(args) > 1 {
+		startKey, err = url.QueryUnescape(args[0])
+		if err != nil {
+			return nil, err
+		}
+		endKey, err = url.QueryUnescape(args[1])
+		if err != nil {
+			return nil, err
+		}
+		args = args[2:]
```
[the remark this answers is not in the document]
Comment (rleungx) [reply server/schedulers/utils.go]: I think there is no much difference between them?

File: server/core/region.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (lhy1024) [reply server/core/region.go]: I just think their names are somewhat similar.

File: server/schedulers/balance_test.go [hunk 5]
```diff
@@ -214,8 +214,8 @@ func (s *testBalanceLeaderSchedulerSuite) TestBalanceLimit(c *C) {
 
 func (s *testBalanceLeaderSchedulerSuite) TestBalanceLeaderScheduleStrategy(c *C) {
 	// Stores:			1    	2    	3    	4
-	// Leader Count:		10    	10    	10    	10
-	// Leader Size :		10000   100    	100    	100
+	// Leader Count:	10    	10    	10    	10
```
Comment (lhy1024) [inline server/schedulers/balance_test.go]: no well format

File: server/schedulers/balance_test.go [hunk 6]
```diff
@@ -335,7 +335,7 @@ func (s *testBalanceLeaderSchedulerSuite) TestBalanceFilter(c *C) {
 }
 
 func (s *testBalanceLeaderSchedulerSuite) TestLeaderWeight(c *C) {
-	// Stores:	1	2	3	4
+	// Stores:		1		2		3		4
```
Comment (lhy1024) [inline server/schedulers/balance_test.go]: ditto

File: server/schedulers/balance_test.go [hunk 7]
```diff
@@ -252,8 +252,8 @@ func (s *testBalanceSpeedSuite) TestTolerantRatio(c *C) {
 func (s *testBalanceLeaderSchedulerSuite) TestBalanceLeaderTolerantRatio(c *C) {
 	// default leader tolerant ratio is 5, when schedule by count
 	// Stores:			1		2    	3    	4
-	// Leader Count:		14->15		10    	10    	10
-	// Leader Size :		100		100    	100    	100
+	// Leader Count:	14->15	10    	10    	10
```
Comment (lhy1024) [inline server/schedulers/balance_test.go]: ditto

File: server/schedulers/balance_test.go [hunk 8]
```diff
@@ -414,6 +414,97 @@ func (s *testBalanceLeaderSchedulerSuite) TestBalanceSelector(c *C) {
...
+	s.tc = mockcluster.NewCluster(opt)
+	s.oc = schedule.NewOperatorController(s.ctx, nil, nil)
+}
+
+func (s *testBalanceLeaderRangeSchedulerSuite) TearDownTest(c *C) {
+	s.cancel()
+}
+
+func (s *testBalanceLeaderRangeSchedulerSuite) TestSingleRangeBalance(c *C) {
+	// Stores:		1		2		3		4
```
Comment (lhy1024) [inline server/schedulers/balance_test.go]: ditto format error and there is similar format in another new test.

File: server/cluster_test.go [hunk 9]
```diff
@@ -833,10 +833,10 @@ func (s *testRegionsInfoSuite) Test(c *C) {
 	}
 
 	for i := uint64(0); i < n; i++ {
-		region := cache.RandLeaderRegion(i, []core.KeyRange{core.NewKeyRange("", "")}, core.HealthRegion())
+		region := cache.RandLeaderRegion(i, []core.KeyRange{core.NewKeyRange("", "")})
```
Comment (lhy1024) [inline server/cluster_test.go]: why to remove `core.HealthRegion()`?

File: server/core/region.go [hunk 10]
```diff
@@ -881,6 +900,10 @@ func DiffRegionKeyInfo(origin *RegionInfo, other *RegionInfo) string {
 	return strings.Join(ret, ", ")
 }
 
+func isInvolved(region *RegionInfo, startKey, endKey []byte) bool {
+	return bytes.Compare(region.GetStartKey(), startKey) >= 0 && (len(endKey) == 0 || (len(region.GetEndKey()) > 0 && bytes.Compare(region.GetEndKey(), endKey) <= 0))
```
Comment (lhy1024) [inline server/core/region.go]: do we need to simplify the logic?

File: server/cluster_test.go (hunk 9 above)
Comment (rleungx) [reply server/cluster_test.go]: Since the function doesn't need this parameter.

File: server/core/region.go (hunk 10 above)
Comment (rleungx) [reply server/core/region.go]: Any better idea?

Comment (rleungx) [reply server/core/region.go]: I think we can make it public when it is necessary.

**Pull request merged**